# mitolib - a library for mtDNA genome data processing

<a href="https://www.java.com/"><img src="http://img.shields.io/badge/language-java-brightgreen.svg" alt="Language" data-canonical-src="http://img.shields.io/badge/language-java-brightgreen.svg" style="max-width:100%;"></a></p>


Current Tools:
- splitter - Split mitochondrial variants and heteroplasmies from <a href="https://mtdna-server.uibk.ac.at/index.html">mtDNA-Server</a>
- splitterRaw - Split mitochondrial variants and heteroplasmies from <a href="https://mtdna-server.uibk.ac.at/index.html">mtDNA-Server</a> raw.txt
- contChecker - Compare mitochondrial profiles from extended report in <a href="http://haplogrep.uibk.ac.at/">HaploGrep 2</a>
- lofreq - Split mitochondrial variants according the VCF file generated with <a href="http://csb5.github.io/lofreq/">LoFreq</a>
- bam2var - naive variant caller
- haplochecker - check for contamination in mtDNA NGS data (BAM file) based on <a href="http://phylotree.org/">Phylotree 17</a> 
- haplochecker2 - check for contamination in <a href="https://mtdna-server.uibk.ac.at/index.html">mtDNA-Server</a>  raw data data (large txt.file) based on <a href="http://phylotree.org/">Phylotree 17</a>   


